### PR TITLE
Bearing displacement

### LINKED ISSUE #70: Don't update fix in bearing if last_fix is within epsilon of new_fix.
Following discussion on https://github.com/gtagency/buzzmobile/pull/65#discussion-diff-84809556R3

If the new_fix is within a small epsilon of the last_fix we kept track of, we don't want to keep this new_fix.

### PR DESCRIPTION
**Pull request by ashlleysam**
Close #70

### CODE REVIEW

**Comment by irapha** [comment]
When this is ready for review, just put the "needs review" label to lmk and I'll look at it

File: buzzmobile/process/bearing/bearing.py [hunk 1]
```diff
@@ -31,7 +31,16 @@ def median(self):
...
+
+
 def bearing(fix):
+    """
+    Determines if the latest position is >= MIN_FIX_DISTANCE from the last
+    position, and if so adds the new bearing to the median filter and
+    publishes the new median.
+    """
     global last_fix
     global med_filter
```
Comment (irapha) [inline buzzmobile/process/bearing/bearing.py]: Idk how these two lines passed code review. Can you please remove these `global` lines?

File: buzzmobile/constants.yaml [hunk 2]
```diff
@@ -13,7 +13,12 @@ pixels_per_m: 80
 # bearing calculation accuracy.
 median_filter_size: 5
 
-# Used for tentacle projection, travel distance is the step between each 
+# Minimum distance a positional fix must be from the previous
+# fix in order to add its bearing to the median filter. Used to
+# improve bearing accuracy at low speeds.
+min_fix_distance: 0.1
```
Comment (irapha) [inline buzzmobile/constants.yaml]: Can you add the unit of this distance to the comment?

Comment (irapha) [reply buzzmobile/constants.yaml]: Also, depending on the units this might be too large a distance, no? If it's radians, then the min distance is enormous for a sphere the size of the earth.

File: buzzmobile/process/bearing/bearing.py [hunk 3]
```diff
@@ -40,13 +49,25 @@ def bearing(fix):
...
 
-        bearing_pub.publish(bearings['med_filter'].median())
+        #Calculates great-circle distance between positions
+        a = (math.pow(math.sin(lat2 - lat1), 2)
+             + math.cos(lat1) * math.cos(lat2)
+             * math.pow(math.sin((lon2 - lon1) / 2), 2))
+        c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
+        distance = EARTH_RADIUS * c
+
+        if distance >= MIN_FIX_DISTANCE:
```
Comment (irapha) [inline buzzmobile/process/bearing/bearing.py]: I'd say we should also save the new fix if we currently don't have any bearing.

Imagine the scenario: we turn BM on, it's stationary. Everything is waiting on the bearing to be published so they can start planning. But bearing is not being published because it's waiting on BM to start moving so it can get a good bearing.

TLDR: when we don't have information, save whatever information we can get.

Comment (irapha) [reply buzzmobile/process/bearing/bearing.py]: And there's probably a way to make the "threshhold" for what is "too low a distance" automatically calculated based on the information we already have. This won't be a part of this issue but it's something to think about. I'll open an issue for it.

File: buzzmobile/process/bearing/bearing.py [hunk 4]
```diff
@@ -40,13 +49,25 @@ def bearing(fix):
...
         lon2 = math.radians(fix.longitude)
-        y = math.sin(lon2 - lon1) * math.cos(lat2)
-        x = math.cos(lat1)*math.sin(lat2) - math.sin(lat1)*math.cos(lat2)*math.cos(lon2 - lon1)
-        angle = math.atan2(y, x)
-        computedBearing = (angle + 2*math.pi) % 2*math.pi
-        bearings['med_filter'].add(computedBearing)
 
-        bearing_pub.publish(bearings['med_filter'].median())
+        #Calculates great-circle distance between positions
+        a = (math.pow(math.sin(lat2 - lat1), 2)
```
Comment (irapha) [inline buzzmobile/process/bearing/bearing.py]: Can you add units to the comment? Is a in meters? radians? km?

Also maybe factor out this great-distance calculation.

File: buzzmobile/process/bearing/bearing.py [hunk 5]
```diff
@@ -40,13 +49,25 @@ def bearing(fix):
...
+        c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
+        distance = EARTH_RADIUS * c
+
+        if distance >= MIN_FIX_DISTANCE:
+            #Calculates forward azimuth between positions
+            y = math.sin(lon2 - lon1) * math.cos(lat2)
+            x = (math.cos(lat1) * math.sin(lat2)
+                 - math.sin(lat1) * math.cos(lat2) * math.cos(lon2 - lon1))
+            angle = math.atan2(y, x)
+            computedBearing = (angle + 2 * math.pi) % (2 * math.pi)
```
Comment (irapha) [inline buzzmobile/process/bearing/bearing.py]: Same for this (units + factor out)?

**Review by irapha — CHANGES_REQUESTED** [review]

File: buzzmobile/process/bearing/bearing.py (hunk 3 above)
Comment (irapha) [reply buzzmobile/process/bearing/bearing.py]: This last part being tracked in #78

File: buzzmobile/process/bearing/bearing.py (hunk 4 above)
Comment (ashlleysam) [reply buzzmobile/process/bearing/bearing.py]: Sorry, what do you mean by factor out the calculation

Comment (irapha) [reply buzzmobile/process/bearing/bearing.py]: I mean creating a new method; something like:

``` python
a = get_great_circle_distance(lat1, lat2, etc)
```

but with a better method name

File: buzzmobile/process/bearing/bearing.py (hunk 3 above)
Comment (ashlleysam) [reply buzzmobile/process/bearing/bearing.py]: So the fix is saved when there's no bearing, but the conditions for it to ~~publish a~~ add a new bearing are that there are at least 2 fixes recorded, and they're at least the delta apart. If a bearing being published is a wait condition should it just ~~publish~~ add 0 radians if it doesn't have data?

Comment (ashlleysam) [reply buzzmobile/process/bearing/bearing.py]: So assuming the delta is not met means that the car is going too slow for bearings to be accurate I would say it should always be ~~publishing a 0~~ adding a 0 to the filter, and publishing the result anyways.

Comment (ashlleysam) [reply buzzmobile/process/bearing/bearing.py]: I totally forgot that a bearing isn't relative, so I actually don't know what to do if there isn't a bearing yet

Comment (irapha) [reply buzzmobile/process/bearing/bearing.py]: I think there's this idea of a good bearing (meets the delta requirement) and a bad bearing (does not meet delta requirement)

Use last fix to update either good or bad bearing.

If there is a last good bearing,  publish that (unless its way too old, but this time part can be implemented when we're fixing #78 )

If there isnt a good bearing, publish the last bearing we were able to get (which can be a bad bearing)

Idk how exactly that translates into implementation, but we want to prioritize good bearings, but not completely fail if we don't have a good bearing. I also don't like the idea of defaulting to 0 rad bearing because our fixes, albeit noisy, might still provide a better bearing than a hardcoded 0 rad.

Comment (irapha) [reply buzzmobile/process/bearing/bearing.py]: If there isnt a bearing, dont publish, but save the fixes so we end up with a bearing at one point (even a bad one, which is better than not having any bearing)

Comment (ashlleysam) [reply buzzmobile/process/bearing/bearing.py]: Alright I understand that. I think we could change the median filter to always pop the oldest value, then when there are no more good bearings just rely on the most recent one.

Comment (irapha) [reply buzzmobile/process/bearing/bearing.py]: Idk if we want that. The purpose of the median filter is to remove noise by outputting the median bearing of the last n bearings we recorded.

Maybe I got confused before and didn't read things properly, but I think the idea behind the issue you're fixing in this PR is to save the `last_fix` (not bearing) if and only if the `last_fix` is greater than a epsilon of the `new_fix`.

Then, following this discussion, if it's _not_ greater than this epsilon _AND_ we have not filled the array in the median filter with bearing calculations, then we should drop the fix.

But we should still be publishing a bearing whenever there is a bearing in the median filter.

Comment (irapha) [reply buzzmobile/process/bearing/bearing.py]: I think my spiel about good and bad bearings was a bad spiel. The median filter is doing that partially.

File: buzzmobile/process/bearing/bearing.py [hunk 6]
```diff
@@ -25,10 +25,10 @@ def median(self):
...
 
 
-#Global Variables
-bearings = {}
-bearings['last_fix'] = None
-bearings['med_filter'] = MedianFilter(rospy.get_param('median_filter_size'))
+# Global variables
+globalVars = {}
+globalVars['last_fix'] = None
+globalVars['med_filter'] = MedianFilter(rospy.get_param('median_filter_size'))
```
Comment (irapha) [inline buzzmobile/process/bearing/bearing.py]: nit: change to `globals`

File: buzzmobile/process/bearing/bearing.py [hunk 7]
```diff
@@ -25,31 +25,71 @@ def median(self):
...
+        if last_fix is not None:
+            distance = get_distance(last_fix, fix)
+            bearing = get_forward_angle(last_fix, fix)
+            med_filter.add(bearing)
+            bearing_pub.publish(med_filter.median())
+            if distance >= MIN_FIX_DISTANCE:
+                last_fix = fix
+        else:
+            last_fix = fix
+    globalVars['last_fix'] = last_fix
```
Comment (irapha) [inline buzzmobile/process/bearing/bearing.py]: nit: use the dict inline. (as in, instead of doing `last_fix = globals['last']` then doing `globals['last'] = last_fix`, just use `globals['last']` everywhere

**Review by irapha — CHANGES_REQUESTED** [review]
two tiny nits but I looked over the actual logic and it looks good. Sorry about the confusion earlier. Your factoring out helped a lot.

File: buzzmobile/process/bearing/bearing.py (hunk 6 above)
Comment (ashlleysam) [reply buzzmobile/process/bearing/bearing.py]: OK cool. I haven't done Python before so please tell me of any stylistic or whatever errors. I understand what the dict was doing when Dip2Dip originally wrote it when he had it kinda badly named as bearings, but when you change the name to `globals` it is highlighted by the IDE? Is `globals` a reserved word or a special dict or something?

**Comment by irapha** [comment]
I dont think `globals` is a reserved word by Python. Just `global` is. So
you can use. Also when we have the pylint working it will yell at you for
stylist errors so you can learn those by the frustration of pylint :)

File: buzzmobile/process/bearing/bearing.py (hunk 7 above)
Comment (ashlleysam) [reply buzzmobile/process/bearing/bearing.py]: Also why do we leave the rsopy publisher outside of the globals dict?

Comment (irapha) [reply buzzmobile/process/bearing/bearing.py]: Because it doesn't get modified. The problem with global variables and ros is that if you have the following:

``` python
a_global

def message_cb(new_msg):
    a_global = new_msg
```

What's actually happening is that the `a_global` inside `message_cb` is not the global variable `a_global`, but a local variable to the method `message_cb`. This is not a problem with ros publishers because:

``` python
pub = rospy.Publisher('msg_name', String, queue_size=1)

def some_method():
    pub.publish('hey')
```

`pub.publish()` would fails if `pub` is a local variable, so python looks up if there's a global named `pub` and finds the publisher we defined.

I don't know why the first problem happens exactly, but it has to do with ros+python interactions because it doesn't happen in vanilla python.

File: buzzmobile/process/bearing/bearing.py [hunk 8]
```diff
@@ -43,8 +43,8 @@ def bearing(fix):
     currently no last_fix
     """
 
-    last_fix = globalVars['last_fix']
-    med_filter = globalVars['med_filter']
+    last_fix = globals['last_fix']
```
Comment (irapha) [inline buzzmobile/process/bearing/bearing.py]: nit: you can also use `globals['last_fix']` inline in the if statement then you can remove this line.

**Review by irapha — CHANGES_REQUESTED** [review]

**Review by irapha — APPROVED** [review]

File: buzzmobile/process/bearing/bearing.py (hunk 7 above)
Comment (joshuamorton) [reply buzzmobile/process/bearing/bearing.py]: rapha no!

Its a scoping rule. If a value is ever assigned inside a function, it is assigned a local varname and a value (see `some_method.__code__.co_varnames`). If it is never assigned to, then a recursive lookup is used to parent (nonlocal) function and finally the global scope.

That's why the result of 

```
a = 5
def func():
    a = 7
func()
print(a)
```

will be 5, and not 7, and also why

```
a = 5
def f():
    b = a
    a = 5
f()
```

will raise UnboundLocalError. Depsite the fact that `a` exists at runtime, at compile time (that is, when the function was compiled), it was created as a local. And that local isn't yet bound in the function.

To fix, use the `global` keyword.

```
a_global
def message_cb(new_msg):
    global a_global
    a_global = new_msg
```

This instructs python to use the global version of `a_global` in this function instead of creating a local.

**Comment by ashlleysam** [comment]
Okay the Python god has spoken and not to rat out Rapha or anything but he
had me remove usages of the global keyword in that method

**Comment by joshuamorton** [comment]
Well, its possible that there's a design issue. Consider using ros as a way to 'store' globals. Think, for example, having a node that subscribes to here, that keeps the last pose, and this node also subscribes to that node. IDK if you can even do that though.

**Comment by irapha** [comment]
Wait if I'm being ratted out I'll point out that I had consulted Josh about this and we both said the use of `global` was kind of icky. That's why we've been doing dicts, because that works.

I believe Josh's "rapha no" comment was bc my explanation was wrong, correct?

We can look up if ros has a better way of storing globals but I'm almost 100% sure the answer is no. Until then I'll say dicts are still the way we're using to store globals, because it works.

**Comment by irapha** [comment]
~~This being said, if we want to move towards using `global a_global`, I'd say open an issue and do it in a separate PR because it will modify way more than `bearing.py`~~

Not doing this. See below

**Comment by joshuamorton** [comment]
Correct, explanation was wrong. Globals are stil icky.

**Comment by ashlleysam** [comment]
Why did my one method low hanging fruit pr have to spawn a whole
application structure discussion -_-

File: buzzmobile/process/bearing/bearing.py [hunk 9]
```diff
@@ -25,31 +25,67 @@ def median(self):
             return (sortedArray[index] + sortedArray[index + 1]) / 2.0
 
 
-#Global Variables
-bearings = {}
-bearings['last_fix'] = None
-bearings['med_filter'] = MedianFilter(rospy.get_param('median_filter_size'))
+# Globals
+globals = {}
```
Comment (irapha) [inline buzzmobile/process/bearing/bearing.py]: Change to `g = {} # globals`

**Review by irapha — CHANGES_REQUESTED** [review]

**Review by irapha — APPROVED** [review]
